Mathlib/Control/Monad/Freer.lean

```diff
@@ -0,0 +1,352 @@
+
+A `Freer f a` is a tree of operations from the type constructor `f`, with leaves of type `a`.
+It has two constructors: `pure` for wrapping a value of type `a`, and `impure` for
+representing an operation from `f` followed by a continuation.
+
+This construction provides a free monad for any type constructor `f`, allowing for composable
+effect descriptions that can be interpreted later. Unlike the traditional Free monad,
+this does not require `f` to be a functor. -/
+inductive Freer.{u, v, w} (f : Type u → Type v) (α : Type w) where
+  | pure : α → Freer f α
+  | impure (ι : Type u) (op : f ι) (cont : ι → Freer f α) : Freer f α
+
+universe u v w
+
+namespace Freer
+
+/-- Map a function over a `Freer` monad. -/
+def map {α β : Type _} (F : Type u → Type v) (f : α → β) : Freer F α → Freer F β :=
+fun FFa =>
+  match FFa with
+  | .pure a => .pure (f a)
+  | .impure ι op cont => .impure ι op (fun z => Freer.map F f (cont z))
+
+instance {F : Type u → Type v} : Functor (Freer F) where
+  map := Freer.map F
+
+instance {F : Type u → Type v} : LawfulFunctor (Freer F) where
+  map_const {α β} := by
+    simp [Functor.mapConst, Functor.map]
+  id_map x := by
+    simp [Functor.map]
+    induction x
+    case pure a => simp [Freer.map]
+    case impure ι op cont ih => simp [Freer.map, ih]
+  comp_map g h x := by
+    simp [Functor.map]
+    induction x
+    case pure a => simp [Freer.map]
+    case impure ι op cont ih => simp [Freer.map, ih]
+
+/-- Bind operation for the `Freer` monad. -/
+def bindFree {a b : Type _} (F : Type u → Type v) (x : Freer F a) (f : a → Freer F b) : Freer F b :=
+  match x with
+  | .pure a => f a
+  | .impure ι op cont => .impure ι op (fun z => bindFree F (cont z) f)
+
+instance {F : Type u → Type v} : Monad (Freer F) where
+  pure := Freer.pure
+  bind := bindFree F
+
+instance FreeLawfulMonad {F : Type u → Type v} : LawfulMonad (Freer F) where
+  bind_pure_comp x y := by
+    simp [Functor.map, bind, pure]; induction y
+    · case pure a => simp [bindFree, map, Pure.pure]
+    · case impure X Fx k ih => simp [bindFree, Freer.map, ih]
+  bind_map f x := by
+    simp [bind, Seq.seq]
+  pure_bind a f := by
+    simp [bind, pure, bindFree]
+  bind_assoc x f g := by
+    simp [bind]; induction x
+    case pure a => simp [bindFree, Freer.map]
+    case impure X Fx k ih => simp [bindFree, Freer.map, ih]
+  seqLeft_eq x y := by
+    simp [Functor.map, SeqLeft.seqLeft, Seq.seq]; induction x
+    case pure a =>
+      simp [bindFree, Freer.map]
+      induction y
+      case pure b => simp [bindFree, Freer.map]
+      case impure X Fy k ih => simp [bindFree, Freer.map, ih]
+    case impure X Fx k ih => simp [bindFree, Freer.map, ih]
+  seqRight_eq x y := by
+    simp [Functor.map, bindFree, Freer.map]; induction x
+    case pure a =>
+      simp [bindFree, Freer.map]
+      induction y
+      case pure b => simp [SeqRight.seqRight, Seq.seq, Functor.map, bindFree, Freer.map]
+      case impure X Fy k ih =>
```

Comment:
I don't think `case` is the preferred style, can you switch to `with` throughout?